src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -0,0 +1,304 @@
+| _ _ _ nil _ := by simp!
+| _ _ _ (cons _ _) _ := by simp [length_reverse_aux, nat.add_succ, nat.succ_add]
+
+@[simp] lemma length_reverse {u v : V} (p : G.walk u v) : p.reverse.length = p.length :=
+by simp [reverse]
+
+/-- The `support` of a walk is the list of vertices it visits in order. -/
+def support : Π {u v : V}, G.walk u v → list V
+| u v nil := [u]
+| u v (cons h p) := u :: p.support
+
+/-- The `edges` of a walk is the list of edges it visits in order. -/
+def edges : Π {u v : V}, G.walk u v → list (sym2 V)
+| u v nil := []
+| u v (@cons _ _ _ x _ h p) := ⟦(u, x)⟧ :: p.edges
+
+@[simp] lemma support_nil {u : V} : (nil : G.walk u u).support = [u] := rfl
+
+@[simp] lemma support_cons {u v w : V} (h : G.adj u v) (p : G.walk v w) :
+  (cons h p).support = u :: p.support := rfl
+
+lemma support_ne_nil {u v : V} (p : G.walk u v) : p.support ≠ [] :=
+by cases p; simp
+
+lemma support_eq {u v : V} (p : G.walk u v) : p.support = u :: p.support.tail :=
+by cases p; simp
+
+@[simp] lemma start_mem_support {u v : V} (p : G.walk u v) : u ∈ p.support :=
+by cases p; simp
+
+@[simp] lemma end_mem_support {u v : V} (p : G.walk u v) : v ∈ p.support :=
+by induction p; simp [*]
+
+lemma chain_adj_support_aux : Π {u v w : V} (h : G.adj u v) (p : G.walk v w),
+  list.chain G.adj u p.support
+| _ _ _ h nil := list.chain.cons h list.chain.nil
+| _ _ _ h (cons h' p) := list.chain.cons h (chain_adj_support_aux h' p)
+
+lemma chain_adj_support : Π {u v : V} (p : G.walk u v), list.chain' G.adj p.support
+| _ _ nil := list.chain.nil
+| _ _ (cons h p) := chain_adj_support_aux h p
+
+/-- Every edge in a walk's edge list is an edge of the graph.
+It is written in this form to avoid unsightly coercions. -/
+lemma edges_subset_edge_set : Π {u v : V} (p : G.walk u v) {e : sym2 V}
+  (h : e ∈ p.edges), e ∈ G.edge_set
+| _ _ (cons h' p') e h := by rcases h with ⟨rfl, h⟩; solve_by_elim
+
+@[simp] lemma edges_nil {u : V} : (nil : G.walk u u).edges = [] := rfl
+
+@[simp] lemma edges_cons {u v w : V} (h : G.adj u v) (p : G.walk v w) :
+  (cons h p).edges = ⟦(u, v)⟧ :: p.edges := rfl
+
+@[simp] lemma length_support {u v : V} (p : G.walk u v) : p.support.length = p.length + 1 :=
+by induction p; simp *
+
+@[simp] lemma length_edges {u v : V} (p : G.walk u v) : p.edges.length = p.length :=
+by induction p; simp *
+
+lemma mem_support_of_mem_edges : Π {t u v w : V} (p : G.walk v w)
+  (he : ⟦(t, u)⟧ ∈ p.edges),
+  t ∈ p.support
+| t u v w (cons h p') he := begin
+  simp only [support_cons, edges_cons, list.mem_cons_iff, quotient.eq] at he ⊢,
+  rcases he with ((he|he)|he),
+  { exact or.inl rfl },
+  { exact or.inr (start_mem_support _) },
+  { exact or.inr (mem_support_of_mem_edges _ he), }
+end
+
+/-- A *trail* is a walk with no repeating edges. -/
+def is_trail {u v : V} (p : G.walk u v) : Prop := p.edges.nodup
+
+/-- A *path* is a trail with no repeating vertices. -/
+structure is_path {u v : V} (p : G.walk u v) : Prop :=
+(to_trail : is_trail p)
```

Comment:
Repeating edges imply repeating vertices, or am I missing something? If so, why is that a structure field and not a lemma?

Reply:
Excellent question, and you are correct that `support.nodup` is sufficient. `is_trail` used to be a structure, and for reasons I've forgotten, I turned it into a plain `def` and folded the `extends` projections into each of the structures.

For dot notation reasons, I'm going to go back to
```
structure is_trail {u v : V} (p : G.walk u v) : Prop :=
(edges_nodup : p.edges.nodup)

structure is_path {u v : V} (p : G.walk u v) extends to_trail : is_trail p : Prop :=
(support_nodup : p.support.nodup)
```
and see what (if anything) isn't as nice. It doesn't matter so much if there are "too many" structure fields so long as we have simpler constructors like
```
lemma is_path.mk' {u v : V} {p : G.walk u v} (h : p.support.nodup) : is_path p
```